### PR TITLE
fix: allow adhoc columns in non-aggregate query

### LINKED ISSUE #21689: [Chart] Can't use custom SQL in chart since 2.0.0 upgrade
We have an issue reported from our user where chart fails when adding a column using custom SQL. The case we have seen is a column with type BIGINT. The user was trying to do some simple calculation like `column_name/6000`, however, it throws an error both on the UI and in the backend server where it says `Expected text or file-like object, got <class 'dict'>`. This should not be chart type specific and we have reproduced this error with both histogram and data table.

This functionality was working as expected before upgrading to 2.0.0, so it is likely to be a regression. And I can reproduce this error in the latest master branch as of 10/03/2022 3:30PM.

#### How to reproduce the bug

1. Go to chart, create a new chart
2. When adding a new column, select custom SQL, and put any valid SQL statement
3. Click on save, and update chart

### Expected results

Chart updated with custom SQL.

### Actual results

Fails due to `Expected text or file-like object, got <class 'dict'>`

#### Screenshots
Google Chrome - Superset - 2022-10-03 at 3 48 29 PM

### Environment

(please complete the following information):

- browser type and version: not browser specific
- superset version: 2.0.0, can be reproduced with latest commit on master branch as of 10/03/2022 3:30PM PST
- python version: 3.8
- node.js version: N/A
- any feature flags active:

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by mayurnewase**
### SUMMARY

fixes: https://github.com/apache/superset/issues/21689
regression from : https://github.com/apache/superset/pull/19454

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -1438,15 +1440,24 @@ def get_sqla_query(  # pylint: disable=too-many-arguments,too-many-locals,too-ma
                 select_exprs.append(outer)
         elif columns:
             for selected in columns:
+                if isinstance(selected, dict):
```
Comment (mayurnewase) [inline superset/connectors/sqla/models.py:1443]: this is the real change, everything else is just setting proper types for columns.

File: superset/superset_typing.py [hunk 2]
```diff
@@ -53,8 +53,8 @@ class AdhocMetric(TypedDict, total=False):
 
 class AdhocColumn(TypedDict, total=False):
     hasCustomLabel: Optional[bool]
-    label: Optional[str]
```
Comment (mayurnewase) [inline superset/superset_typing.py:56]: I couldn't find a way on frontend to set columns without custom label, even if its set as blank it will send empty string.

File: superset/connectors/sqla/models.py (hunk 1 above)
Comment (mayurnewase) [reply superset/connectors/sqla/models.py:1443]: also couldnt' use `utils.is_adhoc_column` as mypy can't figure out types properly.

File: superset/connectors/sqla/models.py [hunk 3]
```diff
@@ -119,6 +119,7 @@
 from superset.superset_typing import (
     AdhocColumn,
     AdhocMetric,
+    Column,
```
Comment (zhaoyongjie) [inline superset/connectors/sqla/models.py:122]:
```
Column as ColumnTyping
```

**Review by zhaoyongjie — COMMENTED** [review]
leave a comment for code convention.

File: superset/connectors/sqla/models.py (hunk 1 above)
Comment (zhaoyongjie) [reply superset/connectors/sqla/models.py:1443]: there is a typing guard at `utils.core.is_adhoc_column` for this case.

File: superset/superset_typing.py (hunk 2 above)
Comment (zhaoyongjie) [reply superset/superset_typing.py:56]: don't add any unrelated changes for the regression bugfix. we could open a new pr for the refactor.

**Comment by mayurnewase** [comment]
thanks for review, will update today.

File: superset/superset_typing.py (hunk 2 above)
Comment (mayurnewase) [reply superset/superset_typing.py]: reverted

File: superset/connectors/sqla/models.py [hunk 4]
```diff
@@ -1438,15 +1438,24 @@ def get_sqla_query(  # pylint: disable=too-many-arguments,too-many-locals,too-ma
                 select_exprs.append(outer)
         elif columns:
             for selected in columns:
+                if is_adhoc_column(selected):
+                    sql = selected.get("sqlExpression") or ""
```
Comment (mayurnewase) [inline superset/connectors/sqla/models.py:1442]: @zhaoyongjie now as `sqlExpression` in `AdhocColumn` is optional, I need to put `or ""` to tell mypy it's a string, is this correct way?

File: superset/connectors/sqla/models.py [hunk 5]
```diff
@@ -1438,15 +1438,24 @@ def get_sqla_query(  # pylint: disable=too-many-arguments,too-many-locals,too-ma
...
-                    selected,
+                    sql,
                     self.database_id,
                     self.schema,
                 )
                 select_exprs.append(
                     columns_by_name[selected].get_sqla_col()
-                    if selected in columns_by_name
-                    else self.make_sqla_column_compatible(literal_column(selected))
+                    if not is_adhoc_column(selected) and selected in columns_by_name
```
Comment (zhaoyongjie) [inline superset/connectors/sqla/models.py:1455]: keeping the same logic as previous
```
if instance(selected, str) and selected in columns_by_name
```

File: tests/integration_tests/charts/data/api_tests.py [hunk 6]
```diff
@@ -764,6 +764,47 @@ def test_with_virtual_table_with_colons_as_datasource(self):
         assert "':xyz:qwerty'" in result["query"]
         assert "':qwerty:'" in result["query"]
 
+    @pytest.mark.usefixtures("load_birth_names_dashboard_with_slices")
+    def test_with_table_columns_without_metrics(self):
```
Comment (zhaoyongjie) [inline tests/integration_tests/charts/data/api_tests.py:768]: These tests don't actually rely on the rest of API so personally, suggestion moving these to the `query_context_test.py` and following this [pattern](https://github.com/apache/superset/blob/aef6217a80d8b7d30120efe6502505edc01eb8d3/tests/integration_tests/query_context_tests.py#L849). This is not required, and you can leave it for later if you are busy.

File: superset/connectors/sqla/models.py (hunk 4 above)
Comment (mayurnewase) [reply superset/connectors/sqla/models.py:1442]: can we directly do selected["sqlExpression"] as we already checked that in is_adhoc_column?
so we get definite str type on _sql and _column_label.

Comment (zhaoyongjie) [reply superset/connectors/sqla/models.py:1442]: sure. the`.get` and square brackets are the same meaning in this case.

File: tests/integration_tests/charts/data/api_tests.py (hunk 6 above)
Comment (mayurnewase) [reply tests/integration_tests/charts/data/api_tests.py:768]: will add those integration tests later.

**Review by zhaoyongjie — APPROVED** [review]
work as expected.

image

**Comment by mayurnewase** [comment]
thanks for review and testing ...

File: superset/connectors/sqla/models.py [hunk 7]
```diff
@@ -119,6 +119,7 @@
 from superset.superset_typing import (
     AdhocColumn,
     AdhocMetric,
+    Column as ColumnTyping,
```
Comment (john-bodley) [inline superset/connectors/sqla/models.py:122]: @mayurnewase out of interest why did you rename this? There doesn't seem to be a name clash per your change.

Comment (mayurnewase) [reply superset/connectors/sqla/models.py:122]: @john-bodley there is import of Column from sqlalchemy on line 50.

**Comment by mayurnewase** [comment]
I think this PR also needs to go in v1.5 as ville tagged corresponding issue for 1.5.
I don't know who is the release manager here, so maybe ping @michael-s-molina

**Comment by michael-s-molina** [comment]
> I think this PR also needs to go in v1.5 as ville tagged corresponding issue for 1.5. I don't know who is the release manager here, so maybe ping @michael-s-molina

I added the v1.5 label. Let's see if it merges cleanly.
